### PR TITLE
Correctly infer argument types for constructor calls

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Signed-off-by: Mihai Budiu <mbudiu@vmware.com>
This PR is extracted from https://github.com/p4lang/p4c/pull/3520.
It fixes a couple of bugs which were discovered using the `list` tests, but it really has nothing to do with lists.
#3520 should be rebased on top of this one.
The attached test used to fail before this fix.

### CODE REVIEW

File: frontends/p4/typeChecking/typeChecker.cpp [hunk 1]
```diff
@@ -980,19 +982,74 @@ TypeInference::checkExternConstructor(const IR::Node* errorPosition,
...
+            newArgs->push_back(arg);
+        }
+        if (!named)
+            ++paramIt;
+    }
+    if (changed)
+        arguments = newArgs;
+    auto objectType = new IR::Type_Extern(ext->srcInfo, ext->name, methodType->typeParameters,
+                                          ext->methods);
+    return std::pair<const IR::Type*, const IR::Vector<IR::Argument>*>(objectType, arguments);
```
Comment (apinski-cavium) [inline frontends/p4/typeChecking/typeChecker.cpp:1052]: Since p4c is now compiling as C++17 you can just do:
`return {objectType, arguments};`
here.

**Review by apinski-cavium — COMMENTED** [review]
Just some simplifications because of C++17. structured binding and initialization of pairs via {} syntax.

File: frontends/p4/typeChecking/typeChecker.cpp [hunk 2]
```diff
@@ -3646,31 +3714,30 @@ const IR::Node* TypeInference::postorder(IR::ConstructorCallExpression* expressi
...
 
     if (simpleType->is<IR::Type_Extern>()) {
-        auto args = checkExternConstructor(expression, simpleType->to<IR::Type_Extern>(),
-                                           expression->arguments);
-        if (args == nullptr)
+        auto typeAndArgs = checkExternConstructor(expression, simpleType->to<IR::Type_Extern>(),
+                                                  expression->arguments);
+        auto contType = typeAndArgs.first;
+        auto newArgs = typeAndArgs.second;
+        if (newArgs == nullptr)
```
Comment (apinski-cavium) [inline frontends/p4/typeChecking/typeChecker.cpp:3721]:
```
auto [contType, newArgs] = checkExternConstructor(expression, simpleType->to<IR::Type_Extern>(),
                                                  expression->arguments);
```
(github did save my comment previously).

**Review by jfingerh — APPROVED** [review]
The test case and expected output looks reasonable to me.  I have not made any attempt to review the changes to the compiler internals, so treat this approval with that qualification.

File: frontends/p4/typeChecking/typeChecker.cpp (hunk 2 above)
Comment (mbudiu-vmw) [reply frontends/p4/typeChecking/typeChecker.cpp:3721]: The bazel build system complains about your proposal:
```
frontends/p4/typeChecking/typeChecker.cpp:1130:14: warning: decomposition declaration only available with -std=c++1z or -std=gnu++1z
         auto [type, newArgs] = checkExternConstructor(decl, et, decl->arguments);
              ^
```
Anyone know where the language level is in the bazel scripts?

**Review by jafingerhut — APPROVED** [review]
Adding review from the account with write access, in case that helps make progress.  Again, I am only able to knowledgeably review the test case and expected output files, which look good.

**Review by rst0git — DISMISSED** [review]
LGTM

**Comment by rst0git** [comment]
It looks like this change breaks the `random.p4` test in the eBPF backend:
```bash
$ p4c-ebpf -Ibackends/ebpf/runtime/p4include --target kernel -o ptf_out/random.c p4testdata/random.p4 --Wdisable=unused --max-ternary-masks 3 -DPORT0=4 -DPORT1=5 -DPORT2=6 -DPORT3=7 -DPORT4=8 -DPORT5=9 -DPSA_RECIRC=2 --xdp2tc=cpumap --arch psa;
p4testdata/random.p4(55): [--Werror=type-error] error: AssignmentStatement
        a.rand.f3 = r3.read();
                  ^
  ---- Actual error:
  /usr/local/share/p4c/p4include/psa.p4(676): Cannot cast implicitly type 'T' to type 'bit<16>'
  extern Random<T> {
                ^
  ---- Originating from:
  p4testdata/random.p4(55): Source expression 'r3.read' produces a result of type 'T' which cannot be assigned to a left-value with type 'bit<16>'
          a.rand.f3 = r3.read();
                      ^^^^^^^^^
  /usr/local/share/p4c/p4include/psa.p4(676)
  extern Random<T> {
                ^
p4testdata/random.p4(64): [--Werror=type-error] error: cast: action argument must be a compile-time constant
        default_action = do_forward((PortId_t) 5);
                                    ^^^^^^^^^^^^
p4testdata/random.p4(70): [--Werror=type-error] error: AssignmentStatement
        a.rand.f2 = r1.read();
                  ^
  ---- Actual error:
  /usr/local/share/p4c/p4include/psa.p4(676): Cannot cast implicitly type 'T' to type 'bit<16>'
  extern Random<T> {
                ^
  ---- Originating from:
  p4testdata/random.p4(70): Source expression 'r1.read' produces a result of type 'T' which cannot be assigned to a left-value with type 'bit<16>'
          a.rand.f2 = r1.read();
                      ^^^^^^^^^
  /usr/local/share/p4c/p4include/psa.p4(676)
  extern Random<T> {
                ^
```

**Comment by mbudiu-vmw** [comment]
@fruffy this fails in cpplint. However, 'make cpplint' no longer runs cpplint. Can we fix that so that is continues to work the same way it did? Also, from the current output logs I cannot understand where the failure is.

**Comment by mbudiu-vmw** [comment]
I see the actual error, but it's not easy to search for it, since the word "error" does not appear close to the problem.

**Comment by fruffy** [comment]
I can add a CMake target. In the meantime, `tools/run-cpplint.sh` should do the same thing.

**Review by rst0git — APPROVED** [review]

**Pull request merged**